python/cudf/cudf/core/dataframe.py

```diff
@@ -4531,130 +4529,322 @@ def _sizeof_fmt(num, size_qualifier):
+
+        >>> df.describe(exclude=[np.number])
+               categorical object
+        count            3      3
+        unique           3      3
+        top              d      a
+        freq             1      1
+
+        Excluding object columns from a ``DataFrame`` description.
+
+        >>> df.describe(exclude=[object])
+               categorical numeric
+        count            3     3.0
+        unique           3    <NA>
+        top              d    <NA>
+        freq             1    <NA>
+        mean          <NA>     2.0
+        std           <NA>     1.0
+        min           <NA>     1.0
+        25%           <NA>     1.5
+        50%           <NA>     2.0
+        75%           <NA>     2.5
+        max           <NA>     3.0
+        """
 
         if not include and not exclude:
-            numeric_data = self.select_dtypes(np.number)
-            output_frame = _create_output_frame(numeric_data, percentiles)
+            default_include = [np.number]
+            if datetime_is_numeric:
+                default_include.append("datetime")
+            data_to_describe = self.select_dtypes(include=default_include)
+            if len(data_to_describe.columns) == 0:
+                data_to_describe = self
 
         elif include == "all":
-            if exclude:
-                raise ValueError("Cannot exclude when include='all'.")
-
-            included_data = self.select_dtypes(np.number)
-            output_frame = _create_output_frame(included_data, percentiles)
-            logging.warning(
-                "Describe does not yet include StringColumns or "
-                "DatetimeColumns."
-            )
+            if exclude is not None:
+                raise ValueError("exclude must be None when include is 'all'")
 
+            data_to_describe = self
         else:
-            if not include:
-                include = np.number
-
-            included_data = self.select_dtypes(
+            data_to_describe = self.select_dtypes(
                 include=include, exclude=exclude
             )
-            if included_data.empty:
+
+            if data_to_describe.empty:
                 raise ValueError("No data of included types.")
-            output_frame = _create_output_frame(included_data, percentiles)
 
-        return output_frame
+        describe_series = [
+            data_to_describe[col].describe(percentiles=percentiles)
+            for col in data_to_describe.columns
+        ]
+        names = []
+        ldesc_indexes = sorted((x.index for x in describe_series), key=len)
+        for idxnames in ldesc_indexes:
+            for name in idxnames.to_pandas():
+                if name not in names:
+                    names.append(name)
+
+        if len(describe_series) == 1:
+            return describe_series[0].to_frame()
+        else:
+            return cudf.concat(
+                [
+                    series.reindex(names, copy=False)
```

Comment:
If we can't just align the names returned from `describe`, should we just allow `cudf.concat` handle aligning the names here instead of calling `reindex` which then I believe will need to check for equality of the indexes.

Reply:
Reindexing is being performed to align with the order of Index values pandas returns i.e., `names`(the end `index`) is always deterministic incase of pandas. If we remove this reindex we will still return correct results just without the same order of index as in Pandas, for example:

```python

df
       b        c
0  hello  0.32340
1  world  0.23432
2  hello  0.00000

>>> df.to_pandas().describe()
            b         c
count       3  3.000000
unique      2       NaN
top     hello       NaN
freq        2       NaN
mean       -1  0.185907
std        -1  0.167047
min        -1  0.000000
25%        -1  0.117160
50%        -1  0.234320
75%        -1  0.278860
max        -1  0.323400

>> df.describe()
            b         c
25%        -1   0.11716
50%        -1   0.23432
75%        -1   0.27886
count       3       3.0
freq        2      <NA>
max        -1    0.3234
mean       -1  0.185907
min        -1       0.0
std        -1  0.167047
top     hello      <NA>
unique      2      <NA>
```

So should we choose to return how it is currently being returned or is non-deterministic order okay?

Reply:
We should choose to return how it is currently being returned, thanks!